src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -17,8 +22,25 @@ using namespace std;
 using namespace ov;
 using namespace ov::op;
 
-bool shared_node_optimization(const shared_ptr<Model>& model,
-                              const unordered_map<Node::type_info_t, bool (*)(const Node*, const Node*)>& rules) {
+using rules_t = unordered_map<Node::type_info_t, bool (*)(const Node*, const Node*)>;
+
+bool type_in_rules(const rules_t& rules, Node::type_info_t& type) {
+    if (rules.count(type))
+        return true;
+    for (const auto& item : rules) {
+        if (type.is_castable(item.first)) {
+            type = item.first;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool greater_type_info(const ov::Node* lhs, const ov::Node* rhs) {
+    return lhs->get_type_info() > rhs->get_type_info();
```

Comment:
it won't work correctly (return false) with
lhs = v10 and rhs = v9

Reply:
I removed this code